solr/core/src/java/org/apache/solr/servlet/CoreContainerProvider.java

```diff
@@ -93,69 +87,69 @@ public class CoreContainerProvider implements ServletContextListener {
   private HttpClient httpClient;
   private SolrMetricManager metricManager;
   private RateLimitManager rateLimitManager;
-  private final CountDownLatch init = new CountDownLatch(1);
   private String registryName;
-  // AFAIK the only reason we need this is to support JettySolrRunner for tests. In tests we might
-  // have multiple CoreContainers in the same JVM, but I *think* that doesn't happen in a real
-  // server.
-  private static final Map<ContextInitializationKey, ServiceHolder> services =
-      Collections.synchronizedMap(new WeakHashMap<>());
-
-  // todo: dependency injection instead, but for now this method and the associated map will have
-  // to suffice.
-  // Note that this relies on ServletContext.equals() not implementing anything significantly
-  // different than Object.equals for its .equals method (I've found no implementation that even
-  // implements it).
-  public static ServiceHolder serviceForContext(ServletContext ctx) throws InterruptedException {
-    ContextInitializationKey key = new ContextInitializationKey(ctx);
-    return services.computeIfAbsent(key, ServiceHolder::new);
+
+  /** Acquires an instance from the context, waiting if necessary. */
+  public static CoreContainerProvider serviceForContext(ServletContext ctx)
+      throws InterruptedException {
+    long startWait = System.nanoTime();
+    synchronized (ctx) {
```

Comment:
Hmm interesting idea, though slightly scary to wait on something jetty owns. Never know when they might also (try to) lock it during server initialization. Also I'm unsure if this would be safe if we ever had another ServletContextListner do the same thing... Worse yet the use of JettySolrRunner for our tests means that we don't have any tests for what start.jar is doing. That dichotomy is a lurking problem, for which there are 2 possible solutions: 1) make jetty startup quicker for tests with https://eclipse.dev/jetty/documentation/jetty-12/operations-guide/index.html#og-quickstart or 2) ditching start.jar and transitioning to custom programatic jetty startup (possibly a variant of JettySolrRunner). As it stands, the tests don't do a good job of covering the true startup process.

Reply:
@epugh was the one who mentioned the jetty quickstart mechanism to me back when I was doing this. I don't know if he's had thoughts since then.

Reply:
> Worse yet the use of JettySolrRunner for our tests means that we don't have any tests for what start.jar is doing.

BATS integration tests is our answer to that.  JettySolrRunner just needs to handle the sequencing realistically (what starts and stops when).  You might have also noticed some bolted-on requirements for tests like debug filter, proxy port, and more, so I don't see JettySolrRunner going away.  It could be simplified further!

I'm not optimistic about that QuickStart thing helping us -- we've already tuned Jetty.  Most notably, our web.xml has `metadata-complete="true"` thus there is no annotation scanning.

Reply:
> Never know when they might also (try to) lock it during server initialization.

I'm not concerned; it's only Jetty and us to worry about.  The code here is resilient to `wait` being notified inexplicably.  There's a loop and it keeps checking for the CoreContainer to show up.  Perhaps it'll log a bit more that it's waiting; it's okay.  In practice I don't think it'll wait at all since you had pointed out in comments that the ServletContextListener is initialized first.

Reply:
Based on the initialization sequence that I'm coming to learn, I suspected that this code doesn't even have to wait and I was right!  I added an assert before ctx.wait and it never tripped.  Ha!

Maybe this is controversial but I'm tempted to remove this code (and the "notify" code elsewhere) and replace it with a lookup that on non-existence throws an IllegalStateException claiming that the ServletContextListener should have done its job by then.  If it didn't, something is seriously wrong.

Reply:
Makes sense to me.  If it doesn't work we can always change back right?